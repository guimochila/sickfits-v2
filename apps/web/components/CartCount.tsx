import styled from '@emotion/styled'
import { TransitionGroup, CSSTransition } from 'react-transition-group'
import Dot from './styled/Dot'

const CartCountAnimation = styled.span`
  position: relative;
  .count {
    display: block;
    position: relative;
    transition: transform 0.4s;
    backface-visibility: hidden;
  }
  .count-enter {
    transform: scale(4) rotateX(0.5turn);
  }
  .count-enter-active {
    transform: rotateX(0);
  }
  .count-exit {
    top: 0;
    position: absolute;
    transform: rotateX(0);
  }
  .count-exit-active {
    transform: scale(4) rotateX(0.5turn);
  }
`

interface ICartCountProps {
  count: number
}

function CartCount({ count }: ICartCountProps) {
  return (
    <CartCountAnimation>
      <TransitionGroup>
        <CSSTransition
          unmountOnExit
          className="count"
          classNames="count"
          key={count}
          timeout={{ enter: 400, exit: 400 }}
        >
          <Dot>{count}</Dot>
        </CSSTransition>
      </TransitionGroup>
    </CartCountAnimation>
  )
}

export default CartCount
