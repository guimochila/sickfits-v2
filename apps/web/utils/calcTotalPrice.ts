import { CartItem } from '../@types/graphql-generated'

function calcTotalPrice(cart: CartItem[]) {
  return cart.reduce((total, cartItem) => {
    if (!cartItem.product) return total

    return total + cartItem.quantity * cartItem.product.price
  }, 0)
}

export default calcTotalPrice
