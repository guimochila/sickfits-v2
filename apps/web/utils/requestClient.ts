import { GraphQLClient } from 'graphql-request'
import { API_ENDPOINT } from '../config'

const gqlClient = new GraphQLClient(API_ENDPOINT, {
  credentials: 'include',
  mode: 'cors',
})

export default gqlClient
