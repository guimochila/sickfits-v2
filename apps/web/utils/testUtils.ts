import casual from 'casual'
import { PAGINATION_QUERY } from '../hooks/useProductsCount'
// seed it so we get consistent results
casual.seed(777)

const fakeItem = (): any => ({
  // __typename: 'Product',
  id: 'abc123',
  price: 5000,
  user: null,
  photo: {
    id: 'abc123',
    altText: 'dogs are best',
    image: {
      publicUrlTransformed: 'http://localhost/dog.jpg',
      id: '5e2a13ff689b2835ae71d1a7',
      filename: 'dog.jpg',
      originalFilename: 'dog.jpg',
      mimetype: 'image/jpeg',
      encoding: '7bit',
      publicUrl: 'http://localhost/dog.jpg',
    },
  },
  name: 'dogs are best',
  description: 'dogs',
  status: 'AVAILABLE',
})

const fakeUser = (overrides = {}) => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  permissions: ['ADMIN'],
  orders: [],
  cart: [],
  ...overrides,
})

const fakeOrderItem = () => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  image: {
    image: `${casual.word}.jpg`,
  },
  name: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
})

const fakeOrder = () => ({
  __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2022-12-11T20:16:13.797Z',
  user: fakeUser(),
})

const fakeCartItem = (overrides = {}) => ({
  __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  product: fakeItem(),
  user: fakeUser(),
  ...overrides,
})

// Fake LocalStorage
class LocalStorageMock {
  store: any

  constructor() {
    this.store = {}
  }

  clear() {
    this.store = {}
  }

  getItem(key: string) {
    return this.store[key] || null
  }

  setItem(key: string, value: any) {
    this.store[key] = value.toString()
  }

  removeItem(key: string) {
    delete this.store[key]
  }
}

function makePaginationMocksFor(length: number) {
  return [
    {
      request: { query: PAGINATION_QUERY },
      result: {
        data: {
          _allProductsMeta: {
            count: length,
          },
          itemsConnection: {
            __typename: 'aggregate',
            aggregate: {
              count: length,
              __typename: 'count',
            },
          },
        },
      },
    },
  ]
}

export {
  makePaginationMocksFor,
  LocalStorageMock,
  fakeItem,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
}
