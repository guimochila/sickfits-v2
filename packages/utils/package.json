{
  "name": "@sickfits/utils",
  "private": true,
  "version": "0.0.1",
  "main": "build/index.js",
  "types": "build/index.d.ts",
  "license": "MIT",
  "author": {
    "name": "Guilherme Scaldelai"
  },
  "devDependencies": {
    "@types/jest": "^28.1.0",
    "babel-jest": "^28.1.0",
    "jest": "^28.1.0",
    "ts-jest": "^28.0.4",
    "typescript": "^4.6.2"
  },
  "scripts": {
    "build": "tsc",
    "test": "NODE_ENV=test jest",
    "test:watch": "NODE_ENV=test jest --watch"
  }
}
