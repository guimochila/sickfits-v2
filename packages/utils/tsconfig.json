{
  "compilerOptions": {
    "target": "es5",
    "module": "commonjs",
    "declaration": true,
    "outDir": "./build",
    "strict": true,
    "skipLibCheck": true
  },
  "include": ["src/**/*"],
  "exclude": ["node_modules", "src/**/*.test.ts"]
}
