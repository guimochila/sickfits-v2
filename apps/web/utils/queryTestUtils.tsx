import { render } from '@testing-library/react'
import { graphql } from 'msw'
import * as React from 'react'
import { QueryClient, QueryClientProvider } from 'react-query'
import { fakeCartItem, fakeItem, fakeUser } from './testUtils'

const product = fakeItem()

export const handlers = [
  // Single Product Query mock
  graphql.query('SINGLE_PRODUCT_QUERY', (req, res, ctx) => {
    const { id } = req.variables

    if (id === product.id) {
      return res(
        ctx.data({
          product,
        }),
      )
    }

    return res(ctx.errors([{ message: 'Item not found!' }]))
  }),

  // Current User Query mock
  graphql.query('CURRENT_USER_QUERY', (req, res, ctx) => {
    const isAuthenticated = sessionStorage.getItem('is-authenticated')

    if (!isAuthenticated) {
      return res(
        ctx.data({
          authenticatedItem: fakeUser(),
        }),
      )
    }

    if (isAuthenticated === 'yesWithCartItems') {
      return res(
        ctx.data({
          authenticatedItem: fakeUser({
            cart: [fakeCartItem()],
          }),
        }),
      )
    }

    return res(
      ctx.data({
        authenticatedItem: fakeUser(),
      }),
    )
  }),

  // Pagination mock
  graphql.query('PAGINATION_QUERY', (req, res, ctx) => {
    return res(
      ctx.data({
        productsCount: 18,
      }),
    )
  }),

  // Signup Mutation mock
  graphql.mutation('SIGNUP_MUTATION', (req, res, ctx) => {
    const { name, email } = req.variables

    return res(
      ctx.data({
        createUser: {
          name,
          email,
        },
      }),
    )
  }),

  // Reset Request Mutation mock
  graphql.mutation('REQUEST_RESET_MUTATION', (req, res, ctx) => {
    const { email } = req.variables

    if (email) {
      return res(
        ctx.data({
          sendUserPasswordResetLink: true,
        }),
      )
    }

    return res(
      ctx.data({
        sendUserPasswordResetLink: null,
      }),
    )
  }),

  // Create Mutation mock
  graphql.mutation('CREATE_PRODUCT_MUTATION', (req, res, ctx) => {
    const { name, price, description } = req.variables

    if (name && price && description) {
      return res(
        ctx.data({
          createProduct: fakeItem(),
        }),
      )
    }

    return res(ctx.errors([{ message: 'Something went wrong' }]))
  }),
]

const createTestQueryClient = () =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  })

export function renderWithClient(ui: React.ReactElement) {
  const testQueryClient = createTestQueryClient()
  const { rerender, ...result } = render(
    <QueryClientProvider client={testQueryClient}>{ui}</QueryClientProvider>,
  )
  return {
    ...result,
    rerender: (rerenderUi: React.ReactElement) =>
      rerender(
        <QueryClientProvider client={testQueryClient}>
          {rerenderUi}
        </QueryClientProvider>,
      ),
  }
}

export function createWrapper() {
  const testQueryClient = createTestQueryClient()
  // eslint-disable-next-line react/display-name
  return ({ children }: { children: React.ReactNode }) => (
    <QueryClientProvider client={testQueryClient}>
      {children}
    </QueryClientProvider>
  )
}
