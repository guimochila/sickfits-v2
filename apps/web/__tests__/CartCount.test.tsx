import { renderWithClient } from '../utils/queryTestUtils'
import CartCount from '../components/CartCount'
import { waitFor } from '@testing-library/react'

describe('<CartCount/>', () => {
  it('renders', () => {
    renderWithClient(<CartCount count={1} />)
  })

  it('Matches Snapshot', () => {
    const { container } = renderWithClient(<CartCount count={1} />)
    expect(container).toMatchSnapshot()
  })

  it('Update via props', async () => {
    const { container, rerender } = renderWithClient(<CartCount count={1} />)
    expect(container.textContent).toBe('1')

    rerender(<CartCount count={2} />)
    await waitFor(() => {
      expect(container.textContent).toBe('2')
    })
  })
})
