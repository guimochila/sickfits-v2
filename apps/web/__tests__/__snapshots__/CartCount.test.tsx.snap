// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartCount/> Matches Snapshot 1`] = `
<div>
  <span
    class="css-1yyllhc"
  >
    <div>
      <div
        class="count css-snu9j6"
      >
        1
      </div>
    </div>
  </span>
</div>
`;
